src/webgpu/api/operation/storage_texture/read_write.spec.ts

```diff
@@ -0,0 +1,326 @@
+
+const kShaderStagesForReadWriteStorageTexture = ['fragment', 'compute'] as const;
+type ShaderStageForReadWriteStorageTexture =
+  (typeof kShaderStagesForReadWriteStorageTexture)[number];
+
+class F extends GPUTest {
+  GetInitialData(storageTexture: GPUTexture): ArrayBuffer {
+    const format = storageTexture.format;
+    const bytesPerBlock = kTextureFormatInfo[format].bytesPerBlock;
+    assert(bytesPerBlock !== undefined);
+
+    const width = storageTexture.width;
+    const height = storageTexture.height;
+    const arrayLayers = storageTexture.depthOrArrayLayers;
+    const initialData = new ArrayBuffer(bytesPerBlock * width * height * arrayLayers);
+    const initialTypedData = this.GetTypedArrayBuffer(initialData, format);
+    for (let z = 0; z < arrayLayers; ++z) {
+      for (let y = 0; y < height; ++y) {
+        for (let x = 0; x < width; ++x) {
+          const index = z * width * height + y * width + x;
+          switch (format) {
+            case 'r32sint':
+              initialTypedData[index] = (index & 1 ? 1 : -1) * (2 * index + 1);
+              break;
+            case 'r32uint':
+              initialTypedData[index] = 2 * index + 1;
+              break;
+            case 'r32float':
+              initialTypedData[index] = (2 * index + 1) / 10.0;
+              break;
+          }
+        }
+      }
+    }
+    return initialData;
+  }
+
+  GetTypedArrayBuffer(arrayBuffer: ArrayBuffer, format: GPUTextureFormat) {
+    switch (format) {
+      case 'r32sint':
+        return new Int32Array(arrayBuffer);
+      case 'r32uint':
+        return new Uint32Array(arrayBuffer);
+      case 'r32float':
+        return new Float32Array(arrayBuffer);
+      default:
+        unreachable();
+        return new Uint8Array(arrayBuffer);
+    }
+  }
+
+  GetExpectedData(
+    shaderStage: ShaderStageForReadWriteStorageTexture,
+    storageTexture: GPUTexture,
+    initialData: ArrayBuffer
+  ): ArrayBuffer {
+    const format = storageTexture.format;
+    const bytesPerBlock = kTextureFormatInfo[format].bytesPerBlock;
+    assert(bytesPerBlock !== undefined);
+
+    const width = storageTexture.width;
+    const height = storageTexture.height;
+    const arrayLayers = storageTexture.depthOrArrayLayers;
+    const bytesPerRowAlignment = align(bytesPerBlock * width, 256);
+    const itemsPerRow = bytesPerRowAlignment / bytesPerBlock;
+
+    const expectedData = new ArrayBuffer(
+      bytesPerRowAlignment * (height * arrayLayers - 1) + bytesPerBlock * width
+    );
+    const expectedTypedData = this.GetTypedArrayBuffer(expectedData, format);
+    const initialTypedData = this.GetTypedArrayBuffer(initialData, format);
+    for (let z = 0; z < arrayLayers; ++z) {
+      for (let y = 0; y < height; ++y) {
+        for (let x = 0; x < width; ++x) {
+          const expectedIndex = z * itemsPerRow * height + y * itemsPerRow + x;
+          switch (shaderStage) {
+            case 'compute': {
```

Comment:
nit: Maybe a small comment saying somewhere that the FS does a * 2 on the same texel, but that the compute shader flips the texture along the diagonal would help readers understand the math here.